Timestamp support + lots of other changes

eliask opened the pull request:
- code is now ok according to shellcheck
- various simplifications and enhancements to code
- hist search now searches for multiple terms with AND clauses.

jcsalterego commented:
@eliask cool! A few preliminary notes:

* I generally like to break commits up instead of one fell swoop. This makes reverts easier, should we need them.
* I'm not 100% committed to shellcheck just yet; I haven't had a chance to install it locally and play around with it. If I do commit to it, I'll want to add it to the CI and maybe make mention of it in the README for contributions.
* The search functionality change is interesting, but I think I have to think through that just a little bit before changing the behavior. For example, I could see using regex (if sqlite supports it) as useful (in line with grep -E).
* I'd rather keep version bumps in master instead of feature branches.

I'll take a closer look when I can but there are some good ideas here!

eliask commented:
I could break this up later if I can find the time and motivation.

One more thing in general that I didn't investigate is the timestamps from bash_history. Maybe it's configurable but for me I have all lines prefixed with their (local, UTC, who knows) timestamps like so:
> $ tail -n4 .bash_history
> #1495730784
> hist
> #1495730787
> cat ~/.bashrc

Maybe historian could have a bi-temporal table for these where the time-of-entry-according-to-bash AND time-of-import are stored separately.

If I can get around to it, I'll probably investigate this further along with more granular commits for the changes I made here.

jcsalterego commented:
There are a lot of good ideas but this has really drifted a bit from master, and there are some design choices that I'm not 100% sold on right now. I'm going to try and take a closer look at the various ideas you've implemented here.

eliask commented:
Yeah, I was just dumping everything there now, though pull requests tend to update automatically.

I think I'll clean it up somewhat later :)

eliask committed:
Simplify subcommand invocation

eliask committed:
Add multiple search argument support for cmd_search

eliask committed:
Abstract sqlite invocations to historian_sqlite()

- Remove useless uses of cat and subshell invocations
- sqlite is now called only from one place, historian_sqlite().

eliask committed:
Normalize sqlite invocation options

Because user .sqliterc may interfere with sqlite, we should explicitly
set certain options to make sure there are no surprises:

.headers off
.mode list

eliask committed:
Remove unused variables

eliask committed:
Revamp zsh and bash import

- Remove now-unneeded magic delimiters, etc.
- zsh and bash parsing now supports exotic multi-line history entries
- history import will now always clean up temporary files even on
  unclean exit.

eliask committed:
Add bi-temporal time support with import metadata

- There now is a metadata table which tracks the import time and any
  other data too as a key=value store.
- Each import run creates a metadata entry and every imported history entry
  is associated with its metadata_id.
- Importing zsh or bash history will not deduplicate commands unless
  they happen at the same time.

eliask committed:
shellcheck-proof test scripts

eliask committed:
Add .historianrc for configurable metadata

Variables defined in .historianrc (interpreted as a shell script)
are copied over to the metadata data on each hist import call.

eliask committed:
Format hist search results in a nicer way

eliask committed:
Tag sqlite timestamps as timestamps

Instead of storing timestamps as opaque integers, tag them as timestamps
so we can easily see a human-readable time when querying the database.

eliask committed:
Print long messages with heredocs

eliask committed:
Import from bash and zsh history simultaneously

- HISTORIAN_SRC is split into HISTORIAN_ZSH and HISTORIAN_BASH
- References to ZSH_EXTENDED_HISTORY is removed and all zsh_history
  files are treated as the extended history format

eliask committed:
Fail in hist import if no input files were found

eliask commented:
So here it is, everything is more or less split off into fine-grained commits.

- zsh and bash support are concurrently supported and imported from.
- zsh and bash history entries are more parsed robustly.
- I should still make the timestamped bash/readline history the default probably
- regarding the failing Travis CI check. I should still check that. Weird, seeing as it works locally for me...

eliask committed:
Invoke bash with #!/usr/bin/env

jcsalterego commented:
Thanks for all these commits! My time for working on this has been diminished to near zero but when I get a chance I'll take a close look, especially the Travis CI error.

jcsalterego commented on hist:
```diff
@@ -1,29 +1,38 @@
-#!/bin/bash
+#! /usr/bin/env bash
```
What's the argument for a space after a shebang? I've never seen this.

I realize on FreeBSD, bash is probably at `/usr/local/bin/bash`. Is `/usr/bin/env` more likely than `/bin/bash`?

eliask replied:
I think I've never encountered an OS where /usr/bin/env does not exist.
FreeBSD doesn't have /bin/bash, and neither does, for example, NixOS.

jcsalterego commented on hist:
```diff
@@ -38,26 +47,86 @@ preflight_check() {
     fi
 }
 
+export_historianrc_variables() {
```
:sparkles: Cool trick with the subshell.

jcsalterego commented on hist:
```diff
@@ -38,26 +47,86 @@ preflight_check() {
+            id INTEGER PRIMARY KEY ASC
+            ,metadata_id INTEGER NOT NULL
+            ,command TEXT NOT NULL
+            ,command_timestamp INTEGER NOT NULL
```
Changing the schema here will break existing databases, which complicates things quite a bit, unless there's some sort of schema versioning or migration support.

eliask replied:
Yeah I schema versioning/migration would be good but I wasn't sure if the complexity warranted it given that it's like version 0.1.x or so now hehe

jcsalterego commented on hist:
```diff
@@ -68,170 +137,290 @@ fail_unless_db_exists() {
+    }
+
+    {
+        if(match($0, /^: ([1-9][0-9]{8,9}):[0-9]+;(.*)$/, a)) {
```
(BSD?) awk complains on my MacOS for this regex; I haven't found out why just yet.

eliask replied:
Hm I thought I tested this with BSD awk but apparently it was GNU awk only :(

jcsalterego commented on hist:
```diff
@@ -68,170 +137,290 @@ fail_unless_db_exists() {
+    # lines like the following for zsh history.
+    # : 1234567890:0;whatever\
+    # : 1234567890:0;whatever\\\\
+    awk '
```
I have mixed feelings about this much awk in a shell script, though I have already put in a decent amount of SQLite in here as well.

eliask replied:
I actually first implemented this in bash but I could not make it fast by any measure. mere 20k lines would take 3 seconds or something on my laptop vs 0.3s or less with awk.

jcsalterego commented on hist:
```diff
@@ -68,170 +137,290 @@ fail_unless_db_exists() {
+    # Test if the line looks like a timestamp (e.g. "#1234567890").
+    #
+    # Assume a timestamp is 9 or 10 digits long.
+    # So we only accept entries made between 1973 and 2286.
```
We better not be using this script in 2287!

jcsalterego commented on hist:
```diff
@@ -1,29 +1,38 @@
+HISTORIAN_ZSH=${HISTORIAN_ZSH-"${HOME}/.zsh_history"} \
 HISTORIAN_DB=${HISTORIAN_DB-"${HOME}/.historian.db"}
 HISTORIAN_SQLITE3=${HISTORIAN_SQLITE3-"$(which sqlite3)"}
-MAGIC="$(echo -e "\x10\x83\xB9\x9F\x34\xB5\x96\x45")" # 0118 999 881 999 119 725 3
```
There was a semi-good reason for using a magic delimiter here, because it's quite hard to get things in the right format for normal SQLite import, so using a magic constant here would almost guarantee lines would never get split up.

eliask replied:
I opted to replace this with consistent use of standard CSV quoted fields. As long as `"` is always escaped with as `""`, nothing else is necessary AFAIK.
This goes together with the bash/zsh command parsing elsewhere.

jcsalterego commented on test/_init.sh:
```diff
@@ -72,8 +70,8 @@ add_failure() {
     local sandbox=$1
-    if [ -n $sandbox ] && [ -n "$(echo $sandbox | grep /tmp)" ]; then
-        rm -rf $sandbox
+    if [[ -n $sandbox ]] && grep -q /tmp <<< "$sandbox"; then
```
Haven't seen a `<<<` herestring before till here -- cool!

eliask replied:
Yeah it's a bash specific extension. Probably for zsh as well.

jcsalterego commented on test/sample.zsh_history:
```diff
@@ -0,0 +1,33 @@
+: 1492370245:0;history-stat
+: 1492370604:50;vi $HISTFILE
+: 1492370674:0;tail $HISTFILEfoo
+: 1496750990:0;{ printf ' }\
```
Great test example. Who types this into their console :P

eliask replied:
I actually first tried to go for 100% accurate command parsing with `bash -n` (which parses syntax and only returns 0 if it's ok and something else otherwise) but when I tested some potential counter-examples, I found this. This does not parse according to `bash -n` but bash nevertheless tries to execute it and stores it in the history file.

jcsalterego reviewed (commented):
Some thoughts!

theY4Kman commented:
FWIW, I just recently migrated to this branch, and used a not-so-super-simple query to migrate. I still rock bash, and my history table was littered with rows containing `#1662176785` unix timestamps in the `command` column; rows with valid commands would find their timestamp in the preceding row, e.g.

| id | metadata\_id | command | command\_timestamp |
| :--- | :--- | :--- | :--- |
| 1 | 4 | xfconf-query -c xfce4-panel -p /plugins/plugin-27/size-max | 1970-01-01 00:00:00 |
| 2 | 4 | #1638675651 | 1970-01-01 00:00:00 |

Sometimes there would be multiple timestamp rows with no commands in between; I imagine these occur when I smash the Enter key at a terminal for some vertical space between commands.

So, what I did was delete (well, rename) my old `history` table, ran `hist` once to import my old `.bash_history` into its newly-created table, and ran this UPDATE query:

```sql
UPDATE history
SET command_timestamp = ts.timestamp
FROM (
    WITH commands AS (
        SELECT id, command
        FROM history
        WHERE NOT command REGEXP '^#\d+'
        ORDER BY id DESC
    )
    SELECT commands.id, commands.command, cast(substr(history.command, 2) AS int) AS timestamp
    FROM commands
    JOIN history ON history.id = commands.id - 1 AND history.command REGEXP '^#\d+'
) AS ts
WHERE history.id = ts.id;
```

Then I followed it up with a DELETE to kill the hanging timestamp `command` rows.

```sql
DELETE FROM history
WHERE command REGEXP '^#\d+';
```

---

On another note: though a `hist /search` was now showing the times of each command, they were being rendered as UNIX timestamps. idk about you, but I find these very hard to decipher :P I changed the search output to:
```sql
coalesce(datetime(command_timestamp, 'unixepoch', 'localtime'), command_timestamp)
```

I also screwed with the formatting a bit, added some colour (as I simply _assume_ the caller is running a 256color term), and sorted by timestamp instead of `id` (well, along with `id`)

```sql
WITH color (reset, bold, grey50, grey58) AS (
    VALUES ('\033[0m', '\033[1m', '\033[38;5;8m', '\033[38;5;246m')
)
SELECT
    grey50 || '#' || grey58 || id
    || grey50 || ' @ ' || grey58 || coalesce(datetime(command_timestamp, 'unixepoch', 'localtime'), command_timestamp) || reset
    || '\n\t' || replace(command, '\n', '\n\t')
FROM history, color
WHERE 1
$(search_terms)
ORDER BY command_timestamp ASC, id DESC;
```

![image](https://user-images.githubusercontent.com/33840/188508254-b15f1ff2-b543-498c-9da2-11b1fa4b02ca.png)

jcsalterego commented:
@theY4Kman thanks for all the feedback and your notes! I have to dust off this project and incorporate some of the changes that @eliask so graciously provided _years_ ago.

jcsalterego commented:
Apologies all - I'm going to finally close this PR.
